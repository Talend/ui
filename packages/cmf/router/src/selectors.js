import match from '@talend/react-cmf/lib/matchPath';

/**
 * getLocation selector return the location from the redux state object.
 * @param {Object} state the redux state
 * @return {Object} location descriptor https://github.com/ReactTraining/react-router/blob/v3/docs/Glossary.md#locationdescriptor
 */
export function getLocation(state) {
	return state.routing.locationBeforeTransitions;
}

/**
 * getPath selector return the path from the redux state object.
 * @param {Object} state the redux state
 * @return {string} path
 */
export function getPath(state) {
	return getLocation(state).pathname;
}

/**
 * matchPath selector provide higher API to match the current path.
 * It give you in one call the composition of matchPath and getPath.
 * @param {Object} state the redux state
 * @param {Object} options matchPath options.
 */
export function matchPath(state, options) {
	return match(getPath(state), options);
}
