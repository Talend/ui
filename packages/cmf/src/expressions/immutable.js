import _get from 'lodash/get';
import Immutable from 'immutable';

/**
 *
 * @param {Object} contextAndPayload is automaticly injected by CMF
 * @param {String} propsPath must be the path of the props you want to getIn
 * @param {String} immutablePath is the path to getIn must be an array
 * @param {Any} defaultValue the value you want to get back if no value found
 * @example
 *    "titleExpression": {
        "id": "Immutable.getIn",
        "args": ["model", ["label"]]
	  },
	  <MyComponent model={myImmutable} />
 */
export function getIn({ context, payload }, propsPath, immutablePath, defaultValue) {
	return _get(payload, propsPath, new Immutable.Map()).getIn(immutablePath, defaultValue);
}

export function get(fullContext, propsPath, id, defaultValue) {
	const immutablePath = id.split('.')
	return getIn(fullContext, propsPath, immutablePath, defaultValue);
}
