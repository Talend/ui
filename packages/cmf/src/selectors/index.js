import * as collections from './collections';
import * as router from './router';
import toJS from './toJS';

export default {
	collections,
	router,
	toJS,
};
