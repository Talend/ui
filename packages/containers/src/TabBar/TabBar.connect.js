import { cmfConnect } from '@talend/react-cmf';
import TabBar, { DEFAULT_STATE } from './TabBar.container';

export default cmfConnect({
	componentId: ownProps => ownProps.componentId || ownProps.id,
	defaultState: DEFAULT_STATE,
})(TabBar);
