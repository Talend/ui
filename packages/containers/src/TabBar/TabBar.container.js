import React from 'react';
import PropTypes from 'prop-types';
import omit from 'lodash/omit';
import Immutable from 'immutable';
import { cmfConnect } from '@talend/react-cmf';
import { TabBar as Component } from '@talend/react-components';

export const DEFAULT_STATE = new Immutable.Map({
	selectedKey: undefined,
});

export const DISPLAY_NAME = 'Container(TabBar)';

class TabBar extends React.Component {
	static displayName = DISPLAY_NAME;

	static propTypes = {
		...cmfConnect.propTypes,
		...Component.propTypes,
		onSelect: PropTypes.func,
		actionCreatorSelect: PropTypes.string,
	};

	constructor(props) {
		super(props);
		this.onSelect = this.onSelect.bind(this);
	}

	onSelect(event, data) {
		if (this.props.onSelect) {
			this.props.onSelect(event, data);
		}
		if (this.props.actionCreatorSelect) {
			this.props.dispatchActionCreator(this.props.actionCreatorSelect, event, { data });
		}
		this.props.setState(() => ({ selectedKey: data.key }));
	}

	render() {
		const state = this.props.state || DEFAULT_STATE;

		const props = Object.assign({}, omit(this.props, cmfConnect.INJECTED_PROPS), {
			onSelect: this.onSelect,
			selectedKey: state.get('selectedKey', undefined) || this.props.selectedKey,
		});

		return <Component {...props} />;
	}
}

export default TabBar;
