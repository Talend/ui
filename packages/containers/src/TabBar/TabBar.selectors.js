import { DEFAULT_STATE, DISPLAY_NAME } from './TabBar.container';

/**
 * Selector on the state from the tabbar id given.
 * @param {object} state
 * @param {string} idComponent
 */
export function getComponentState(state, idComponent) {
	return state.cmf.components.getIn([DISPLAY_NAME, idComponent], DEFAULT_STATE);
}

/**
 * Return the selectedKey attr from the state of tabbar id given.
 * @param {object} state
 * @param {string} idComponent
 */
export function getSelectedKey(state, idComponent) {
	return getComponentState(state, idComponent).get('selectedKey', undefined);
}
