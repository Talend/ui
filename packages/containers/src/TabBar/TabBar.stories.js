import React from 'react';
import TabBar from '.';

const view = {
	items: [
		{
			id: 1,
			label: 'School',
			key: 'school',
		},
		{
			id: 2,
			label: 'Office',
			key: 'office',
		},
		{
			id: 3,
			label: 'Hospital',
			key: 'hospital',
		},
	],
	selectedKey: 'office',
};

export default {
	title: 'TabBar',
};

export const Default = () => <TabBar {...view} />;
