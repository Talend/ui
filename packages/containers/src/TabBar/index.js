import TabBar from './TabBar.connect';
import { getSelectedKey, getComponentState } from './TabBar.selectors';

TabBar.selectors = {
	getSelectedKey,
	getComponentState,
};

export default TabBar;
