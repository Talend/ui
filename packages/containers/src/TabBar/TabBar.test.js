import React from 'react';
import { shallow, mount } from 'enzyme';
import { Map } from 'immutable';
import Container, { DEFAULT_STATE, DISPLAY_NAME } from './TabBar.container';
import Connected from './TabBar.connect';
import { getComponentState, getSelectedKey } from './TabBar.selectors';

describe('TabBar connected', () => {
	it('should connect TabBar', () => {
		expect(Connected.displayName).toBe(`Connect(CMF(${Container.displayName}))`);
		expect(Connected.WrappedComponent).toBe(Container);
	});
});

describe('TabBar Container', () => {
	it('should render', () => {
		const props = {
			items: [
				{ id: 'hello', label: 'Hello', key: 'awesomeKey' },
				{ id: 'world', label: 'WorLd', key: 'specialKey' },
			],
			selectedKey: 'hello',
		};
		const wrapper = shallow(<Container {...props} />);
		expect(wrapper.getElement()).toMatchSnapshot();
	});
	it('should setState when switch tab', () => {
		let state;
		const props = {
			items: [
				{ id: 'some', label: 'someTab', key: 'someTab' },
				{ id: 'someother', label: 'someOtherTab', key: 'someOtherTab' },
			],
			state: Map({ selectedKey: 'someTab' }),
			setState: jest.fn(fn => (state = fn())),
		};

		const wrapper = mount(<Container {...props} />);
		wrapper
			.find('button')
			.at(1)
			.simulate('click');

		expect(props.setState).toHaveBeenCalled();
		expect(state.selectedKey).toEqual('someOtherTab');
	});
});

describe('TabBar selectors', () => {
	let mockState;
	const componentState = Map({ selectedKey: 'hello' });
	beforeEach(() => {
		mockState = {
			cmf: { components: Map({ [DISPLAY_NAME]: Map({ thisTabBar: componentState }) }) },
		};
	});

	it('should return the state', () => {
		expect(getComponentState(mockState, 'thisTabBar')).toEqual(componentState);
	});
	it('should return the default state', () => {
		expect(getComponentState(mockState, 'idNotExist')).toEqual(DEFAULT_STATE);
	});
	it('should return selectedKey', () => {
		expect(getSelectedKey(mockState, 'thisTabBar')).toEqual('hello');
	});
	it('should return default value if pass wrong componentId', () => {
		expect(getSelectedKey(mockState, 'idNotExist')).toBe(undefined);
	});
});
