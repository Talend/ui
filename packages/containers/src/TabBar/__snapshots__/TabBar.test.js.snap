// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabBar Container should render 1`] = `
<TabBar
  items={
    Array [
      Object {
        "id": "hello",
        "key": "awesomeKey",
        "label": "Hello",
      },
      Object {
        "id": "world",
        "key": "specialKey",
        "label": "WorLd",
      },
    ]
  }
  onSelect={[Function]}
  selectedKey="hello"
/>
`;
